pydantic_ai_slim/pydantic_ai/toolsets/searchable.py

```diff
@@ -0,0 +1,180 @@
+import re
+from dataclasses import dataclass, field, replace
+from typing import Any
+
+from .._run_context import AgentDepsT, RunContext
+from ..tools import Tool, ToolDefinition
+from .abstract import ToolsetTool
+from .wrapper import WrapperToolset
+
+_TOOL_SEARCH_METADATA_KEY = '__tool_search__'
+"""ToolDefinition.metadata key to store runtime metadata."""
+
+
+@dataclass
+class SearchableToolset(WrapperToolset[AgentDepsT]):
+    """A toolset that implements tool search and deferred tool loading."""
+
+    _active_tool_names: set[str] = field(default_factory=set)
+    """Tracks activated tool names."""
+
+    async def get_tools(self, ctx: RunContext[AgentDepsT]) -> dict[str, ToolsetTool[AgentDepsT]]:
+        tools = await self.wrapped.get_tools(ctx)
+
+        # No need to add a search tool if there are no defer_loading tools.
+        if not any(t.tool_def.defer_loading for t in tools.values()):
+            return tools
+
+        # If any nested SearchableToolset instances have added their own search tools, drop those so that only one
+        # search tool is exposed to the model.
+        tools = {name: tool for name, tool in tools.items() if not is_search_tool(tool.tool_def)}
+
+        all_tools: dict[str, ToolsetTool[AgentDepsT]] = {}
+
+        for tool_name, tool in tools.items():
+            defer = tool.tool_def.defer_loading
+            all_tools[tool_name] = _SearchToolsetToolWrapper[AgentDepsT](tool, self) if defer else tool
+
+        search_tool, search_toolset_tool = _search_tool(toolset=self)
+
+        # TODO how to handle this error, or should we automatically disambiguate this name?
```

Comment:
Need to look is it idiomatic in the codebase to generate unique tool names, probably this needs to be done deterministically.